util/route/route.go

```diff
@@ -0,0 +1,98 @@
+package route
+
+import (
+	"net/http"
+	"sync"
+
+	"github.com/julienschmidt/httprouter"
+	"golang.org/x/net/context"
+)
+
+var (
+	mtx   = sync.RWMutex{}
+	ctxts = map[*http.Request]context.Context{}
+)
+
+// Context returns the context for the request.
+func Context(r *http.Request) context.Context {
+	mtx.RLock()
+	defer mtx.RUnlock()
+	return ctxts[r]
+}
+
+type param string
+
+// Param returns param p for the context.
+func Param(ctx context.Context, p string) string {
+	return ctx.Value(param(p)).(string)
+}
+
+// handle turns a Handle into httprouter.Handle
+func handle(h http.HandlerFunc) httprouter.Handle {
+	return func(w http.ResponseWriter, r *http.Request, params httprouter.Params) {
+		ctx, cancel := context.WithCancel(context.Background())
+		defer cancel()
+
+		for _, p := range params {
```

Comment:
In my experience, since the handler allocates everything only once, the params slice is reused by repeated (or even concurrent) calls of the same handler. Accessing params is therefore a race. In the pushgateway, I needed to protect params with a mutex. You should definitely test this with the race detector.